server/backend/api/app/routers/router_guards.py

```diff
@@ -184,31 +196,71 @@ def authorize_by_application_role(
-            detail={
-                "code": ERROR_INVALID_ROLE_ID,
-                "description": f"Requester has no appropriate role.",
-            },
-            headers={"WWW-Authenticate": "Bearer"},
-        )
-
     # Delegate to this for main check.
     authorize_by_app_id(
         application_id=role.application_id, db=db, claims=claims, requester=requester
     )
     return role
 
 
-async def internal_only_action(requester: Requester = Depends(get_current_requester)):
-    if requester.user_type_code is not UserType.IDIR:
-        raise external_user_prohibited_exception
+async def authorize_by_privilege(
+    request: Request,
+    role: FamRole = Depends(get_request_role_from_id),
+    db: Session = Depends(database.get_db),
+    claims: dict = Depends(validate_token),
+    requester: Requester = Depends(get_current_requester),
+):
+    """
+    This authorize_by_privilege method is used for checking if the requester has the privilege to grant/remove access of the role.
+    :param role: for remove access, it is the concrete role get by user_role_xref_id in the request params;
+                 for grant access, it is the concrete role, or the parent role of an abstract role get by role_id in the request params
+    """
+    requester_is_app_admin = is_app_admin(role.application_id, db, claims)
+
+    # if requester is not application admin, check if has the privilege to grant/remove access of the role
+    if not requester_is_app_admin:
+        # for remove access, role is what we get above from Depends(get_request_role_from_id)
+        # for grant access, if the request params has forest_client_number, need to get the child role
+        # the child role should already exist when the delegated admin has been created
+        if "user_role_xref_id" not in request.path_params:
+            rbody = await request.json()
+            forest_client_number = rbody.get("forest_client_number")
+            if (
+                forest_client_number
+                and role.role_type_code == RoleType.ROLE_TYPE_ABSTRACT
+            ):
+                parent_role = role
+                forest_client_role_name = (
+                    crud_user_role.construct_forest_client_role_name(
+                        parent_role.role_name, forest_client_number
+                    )
+                )
+                role = crud_role.get_role_by_role_name_and_app_id(
+                    db, forest_client_role_name, parent_role.application_id
+                )  # when role is None, means the role is not created for this delegated admin
+
+        requester_has_privilege = False
+        if role:
+            requester_has_privilege = (
+                crud_access_control_privilege.has_privilege_by_role_id(
+                    db, requester.user_id, role.role_id
+                )
+            )
+        # if user has no privilege of the role, throw permission error
+        if not requester_has_privilege:
```

Comment:
Just curious, maybe I am still confused.
Does this need to be `if (role is None) or not requester_has_privilege:`?

Reply:
um... yeah, but feel no need to add role is None here? cause if role is None, we won't be able to get role_id, so the requester_has_privilege will be None?
<img width="679" alt="Screen Shot 2024-04-11 at 4 03 36 PM" src="https://github.com/bcgov/nr-forests-access-management/assets/23131735/3ab10cac-bee0-4ac3-9930-dcfccfa221d5">
or it can be like the following, do you prefer this way maybe?
```
if not role or not crud_access_control_privilege.has_privilege_by_role_id(
    db, requester.user_id, role.role_id
):
```

Reply:
Ah I see, I didn't follow it clearly.